queries/cpp/injections.scm

```diff
@@ -1,3 +1,3 @@
-(preproc_arg) @cpp
+; inherits: c
```

Comment:
Inherits c is problematic as we don't want the inject `@c`. This is why before everything was repeated

Reply:
Should we factor out the c/c++ common stuff into something like `queries/c_common`?

Reply:
Or alternatively add a predicate for the `@c` injection.

Reply:
We would need a `@self` injection. That would simplify a lot of things

Reply:
> Inherits c is problematic as we don't want the inject `@c`.

The `@cpp` injection overrides that but it might still have both injections.

Reply:
does it overwrite and only have one injection (you can check the language tree Lua table)? I'm ok if you tested that it works. The query ordering has changed since when the files were created.

Reply:
It looks like the `@c` injection remains but `@cpp` takes priority.

Reply:
But does Neovim inject both languages?

Reply:
Yes it does.

#### tree.lua

```lua
LanguageTree = require('vim.treesitter.languagetree')
tree = LanguageTree.new([[#define FOO (bar)]], 'cpp')
tree:parse()
vim.pretty_print(tree:_get_injections())
vim.pretty_print(tree:language_for_range({0, 11, 0, 17})._lang)
```

#### test command

```bash
nvim --headless -c 'so tree.lua' +q
```

#### output

```bash
{
  c = { { { 0, 11, 0, 17 } } },
  cpp = { { { 0, 11, 0, 17 } } }
}
"cpp"
```

Reply:
that's not good. We only need one the two.

Reply:
So I guess there are 2 solutions:

1. Move the `printf` injections to a `c_printf` folder and inherit from that.
2. Convert this PR to a draft and wait for a `@self` injection.

Reply:
> Move the printf injections to a c_printf folder and inherit from that.

Not a good solution, rather we would have a common file for all of C injections

> Convert this PR to a draft and wait for a @self injection.

I think `@self` would probably a more elegant solution. Do you want to create a PR to nvim (or else I can do tonight)? We can go for the former solution in the meantime in case `@self` PR takes longer than expected.

About the performance concerns, I think it would be good to add functionality in to tree-sitter to selectively disable queries. So that queries for languages without parsers can be disabled (or go that wiki approach, though I think printf is fairly common so that it could be a default experience).

Reply:
> Do you want to create a PR to nvim (or else I can do tonight)?

I could try but I'm not familiar with the codebase.

Reply:
How's the status of this? It'd be nice to add, but I think it should be `@character.special` or `@string.special`for the format pieces

Reply:
Blocked by neovim/neovim#20712 for the time being.